### PR TITLE
Device Registry Implementation backed with infinispan caches

### PR DESCRIPTION
**Pull request by jbtrystram**
This is an implementation of the device registry APIs based on Infinispan caches. It currently used embedded in-momory caches, and it should be able to be distributed as the number of instances of the DR are increased.

Signed-off-by: Jean-Baptiste Trystram <jbtrystram@redhat.com>

### CODE REVIEW

File: services/infinispan-device-registry/src/main/java/org/eclipse/hono/deviceregistry/CacheCredentialConfigProperties.java [hunk 1]
```diff
@@ -0,0 +1,20 @@
+/*******************************************************************************
+ * Copyright (c) 2016, 2018 Contributors to the Eclipse Foundation
```
Comment (sophokles73) [inline services/infinispan-device-registry/src/main/java/org/eclipse/hono/deviceregistry/CacheCredentialConfigProperties.java]: in new files, use the current year (2019) only

File: services/infinispan-device-registry/src/main/java/org/eclipse/hono/deviceregistry/CacheCredentialService.java [hunk 2]
```diff
@@ -0,0 +1,246 @@
+/*******************************************************************************
+ * Copyright (c) 2016, 2018 Contributors to the Eclipse Foundation
```
Comment (sophokles73) [inline services/infinispan-device-registry/src/main/java/org/eclipse/hono/deviceregistry/CacheCredentialService.java]: 2019

File: services/infinispan-device-registry/src/main/java/org/eclipse/hono/deviceregistry/CacheRegistrationConfigProperties.java [hunk 3]
```diff
@@ -0,0 +1,20 @@
+/*******************************************************************************
+ * Copyright (c) 2016, 2018 Contributors to the Eclipse Foundation
```
Comment (sophokles73) [inline services/infinispan-device-registry/src/main/java/org/eclipse/hono/deviceregistry/CacheRegistrationConfigProperties.java]: 2019

File: services/infinispan-device-registry/src/main/java/org/eclipse/hono/deviceregistry/CacheRegistrationService.java [hunk 4]
```diff
@@ -0,0 +1,106 @@
+/*******************************************************************************
+ * Copyright (c) 2016, 2018 Contributors to the Eclipse Foundation
```
Comment (sophokles73) [inline services/infinispan-device-registry/src/main/java/org/eclipse/hono/deviceregistry/CacheRegistrationService.java]: 2019

File: services/infinispan-device-registry/src/main/java/org/eclipse/hono/deviceregistry/CacheTenantConfigProperties.java [hunk 5]
```diff
@@ -0,0 +1,20 @@
+/*******************************************************************************
+ * Copyright (c) 2016, 2018 Contributors to the Eclipse Foundation
```
Comment (sophokles73) [inline services/infinispan-device-registry/src/main/java/org/eclipse/hono/deviceregistry/CacheTenantConfigProperties.java]: 2019

File: services/infinispan-device-registry/src/main/java/org/eclipse/hono/deviceregistry/CacheTenantService.java [hunk 6]
```diff
@@ -0,0 +1,178 @@
+/*******************************************************************************
+ * Copyright (c) 2016, 2018 Contributors to the Eclipse Foundation
```
Comment (sophokles73) [inline services/infinispan-device-registry/src/main/java/org/eclipse/hono/deviceregistry/CacheTenantService.java]: 2019

File: services/infinispan-device-registry/src/main/java/org/eclipse/hono/deviceregistry/CacheTenantService.java [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+                .from(RegistryTenantObject.class)
+                .having("trustedCa")
+                .contains(subjectDnName).build();
+
+        final List<RegistryTenantObject> matches = query.list();
+
+        // TODO make a difference between not found and conflict?
+        if (matches.size() != 1){
+            return null;
+        }else {
```
Comment (sophokles73) [inline services/infinispan-device-registry/src/main/java/org/eclipse/hono/deviceregistry/CacheTenantService.java]: whitespace

File: services/infinispan-device-registry/src/main/java/org/eclipse/hono/deviceregistry/CredentialsKey.java [hunk 8]
```diff
@@ -0,0 +1,60 @@
+/*******************************************************************************
+ * Copyright (c) 2016, 2018 Contributors to the Eclipse Foundation
```
Comment (sophokles73) [inline services/infinispan-device-registry/src/main/java/org/eclipse/hono/deviceregistry/CredentialsKey.java]: 2019

File: services/infinispan-device-registry/src/main/java/org/eclipse/hono/deviceregistry/InfinispanRegistry.java [hunk 9]
```diff
@@ -0,0 +1,41 @@
+/*******************************************************************************
+ * Copyright (c) 2016, 2018 Contributors to the Eclipse Foundation
```
Comment (sophokles73) [inline services/infinispan-device-registry/src/main/java/org/eclipse/hono/deviceregistry/InfinispanRegistry.java]: 2019

File: services/infinispan-device-registry/src/main/java/org/eclipse/hono/deviceregistry/InfinispanRegistryConfig.java [hunk 10]
```diff
@@ -0,0 +1,48 @@
+/*******************************************************************************
+ * Copyright (c) 2016, 2018 Contributors to the Eclipse Foundation
```
Comment (sophokles73) [inline services/infinispan-device-registry/src/main/java/org/eclipse/hono/deviceregistry/InfinispanRegistryConfig.java]: 2019

File: services/infinispan-device-registry/src/main/java/org/eclipse/hono/deviceregistry/RegistrationKey.java [hunk 11]
```diff
@@ -0,0 +1,56 @@
+/*******************************************************************************
+ * Copyright (c) 2016, 2018 Contributors to the Eclipse Foundation
```
Comment (sophokles73) [inline services/infinispan-device-registry/src/main/java/org/eclipse/hono/deviceregistry/RegistrationKey.java]: 2019

File: services/infinispan-device-registry/src/main/java/org/eclipse/hono/deviceregistry/RegistryCredentialObject.java [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+
+/**
+ * A custom class to be used as value in the backend key-value storage.
+ * This store credentials details.
+ *
+ *  See {@link org.eclipse.hono.deviceregistry.CacheTenantService CacheTenantService} class.
+ */
+public class RegistryCredentialObject {
+
+    private final String tenantId;
```
Comment (sophokles73) [inline services/infinispan-device-registry/src/main/java/org/eclipse/hono/deviceregistry/RegistryCredentialObject.java]: what is this good for if it is not used anywhere?

File: services/infinispan-device-registry/src/main/java/org/eclipse/hono/deviceregistry/RegistryCredentialObject.java [hunk 13]
```diff
@@ -0,0 +1,50 @@
+/*******************************************************************************
+ * Copyright (c) 2016, 2018 Contributors to the Eclipse Foundation
```
Comment (sophokles73) [inline services/infinispan-device-registry/src/main/java/org/eclipse/hono/deviceregistry/RegistryCredentialObject.java]: 2019

File: services/infinispan-device-registry/src/main/java/org/eclipse/hono/deviceregistry/RegistryTenantObject.java [hunk 14]
```diff
@@ -0,0 +1,52 @@
+/*******************************************************************************
+ * Copyright (c) 2016, 2018 Contributors to the Eclipse Foundation
```
Comment (sophokles73) [inline services/infinispan-device-registry/src/main/java/org/eclipse/hono/deviceregistry/RegistryTenantObject.java]: 2019

File: services/infinispan-device-registry/src/main/resources/application.yml [hunk 15]
```diff
@@ -0,0 +1,43 @@
...
+    file: /src/main/resources/cache-configuration.xml
+spring:
+  jmx:
+    enabled: false
+management:
+  metrics:
+    enabled: false
+    export:
+      graphite:
+        enabled: false
```
Comment (sophokles73) [inline services/infinispan-device-registry/src/main/resources/application.yml]: missing EOL

File: services/infinispan-device-registry/src/main/resources/logback.xml [hunk 16]
```diff
@@ -0,0 +1,26 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!--
+    Copyright (c) 2016, 2017 Eurotech and/or its affiliates and others
```
Comment (sophokles73) [inline services/infinispan-device-registry/src/main/resources/logback.xml]: let's hope not
2019

File: services/infinispan-device-registry/src/main/resources/logback.xml [hunk 17]
```diff
@@ -0,0 +1,26 @@
...
+
+    All rights reserved. This program and the accompanying materials
+    are made available under the terms of the Eclipse Public License v1.0
+    which accompanies this distribution, and is available at
+    http://www.eclipse.org/legal/epl-v10.html
+
+    Contributors:
+        Eurotech - initial API and implementation
```
Comment (sophokles73) [inline services/infinispan-device-registry/src/main/resources/logback.xml]: remove

File: services/infinispan-device-registry/src/main/resources/logback.xml [hunk 18]
```diff
@@ -0,0 +1,26 @@
...
+            <pattern>%d{HH:mm:ss.SSS} [%thread] %-5level %logger{36} - %msg%n</pattern>
+        </encoder>
+    </appender>
+
+    <logger name="org.eclipse.hono" level="DEBUG"/>
+
+    <root level="INFO">
+        <appender-ref ref="STDOUT" />
+    </root>
+</configuration>
```
Comment (sophokles73) [inline services/infinispan-device-registry/src/main/resources/logback.xml]: missing EOL

File: services/infinispan-device-registry/src/test/java/org/eclipse/hono/deviceregistry/CacheCredentialsServiceTest.java [hunk 19]
```diff
@@ -0,0 +1,48 @@
+/*******************************************************************************
+ * Copyright (c) 2016, 2018 Contributors to the Eclipse Foundation
```
Comment (sophokles73) [inline services/infinispan-device-registry/src/test/java/org/eclipse/hono/deviceregistry/CacheCredentialsServiceTest.java]: 2019

File: services/infinispan-device-registry/src/test/java/org/eclipse/hono/deviceregistry/CacheRegistrationServiceTest.java [hunk 20]
```diff
@@ -0,0 +1,47 @@
+/*******************************************************************************
+ * Copyright (c) 2016, 2018 Contributors to the Eclipse Foundation
```
Comment (sophokles73) [inline services/infinispan-device-registry/src/test/java/org/eclipse/hono/deviceregistry/CacheRegistrationServiceTest.java]: 2019

File: services/infinispan-device-registry/src/test/java/org/eclipse/hono/deviceregistry/CacheTenantServiceTest.java [hunk 21]
```diff
@@ -0,0 +1,45 @@
+/*******************************************************************************
+ * Copyright (c) 2016, 2018 Contributors to the Eclipse Foundation
```
Comment (sophokles73) [inline services/infinispan-device-registry/src/test/java/org/eclipse/hono/deviceregistry/CacheTenantServiceTest.java]: 2019

File: services/infinispan-device-registry/src/test/resources/logback.xml [hunk 22]
```diff
@@ -0,0 +1,26 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!--
+    Copyright (c) 2016, 2017 Eurotech and/or its affiliates and others
```
Comment (sophokles73) [inline services/infinispan-device-registry/src/test/resources/logback.xml]: 2019
Eurotech?

File: services/infinispan-device-registry/src/test/resources/logback.xml [hunk 23]
```diff
@@ -0,0 +1,26 @@
...
+
+    All rights reserved. This program and the accompanying materials
+    are made available under the terms of the Eclipse Public License v1.0
+    which accompanies this distribution, and is available at
+    http://www.eclipse.org/legal/epl-v10.html
+
+    Contributors:
+        Eurotech - initial API and implementation
```
Comment (sophokles73) [inline services/infinispan-device-registry/src/test/resources/logback.xml]: remove

File: services/infinispan-device-registry/src/test/resources/logback.xml [hunk 24]
```diff
@@ -0,0 +1,26 @@
...
+            <pattern>%d{HH:mm:ss.SSS} [%thread] %-5level %logger{36} - %msg%n</pattern>
+        </encoder>
+    </appender>
+
+    <logger name="org.eclipse.hono" level="DEBUG"/>
+
+    <root level="INFO">
+        <appender-ref ref="STDOUT" />
+    </root>
+</configuration>
```
Comment (sophokles73) [inline services/infinispan-device-registry/src/test/resources/logback.xml]: missing EOL

File: services/infinispan-device-registry/src/main/java/org/eclipse/hono/deviceregistry/RegistryCredentialObject.java (hunk 12 above)
Comment (jbtrystram) [reply services/infinispan-device-registry/src/main/java/org/eclipse/hono/deviceregistry/RegistryCredentialObject.java]: When doing a batch request (i.e. `getAll` or `removeAll`) the value is used as a field to do a query against the cache.
See `queryAllCredentialsForDevice(String, String)` in `CredentialCache`

File: services/infinispan-device-registry/src/test/resources/logback.xml [hunk 25]
```diff
@@ -0,0 +1,27 @@
...
```
Comment (sophokles73) [inline services/infinispan-device-registry/src/test/resources/logback.xml]: missing EOL

File: services/infinispan-device-registry/src/main/resources/application.yml [hunk 26]
```diff
@@ -0,0 +1,43 @@
+hono:
+  app:
+    type: infinispan
```
Comment (sophokles73) [inline services/infinispan-device-registry/src/main/resources/application.yml]: what is this property good for? Can it have another value?

File: services/infinispan-device-registry/src/main/java/org/eclipse/hono/registry/infinispan/InfinispanRegistryConfig.java [hunk 27]
```diff
@@ -0,0 +1,49 @@
...
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *******************************************************************************/
+
+package org.eclipse.hono.registry.infinispan;
+
+import org.eclipse.hono.deviceregistry.ApplicationConfig;
```
Comment (sophokles73) [inline services/infinispan-device-registry/src/main/java/org/eclipse/hono/registry/infinispan/InfinispanRegistryConfig.java]: IMHO we should not have a dependency on another registry implementation.

File: services/infinispan-device-registry/src/main/java/org/eclipse/hono/registry/infinispan/InfinispanRegistry.java [hunk 28]
```diff
@@ -0,0 +1,42 @@
...
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *******************************************************************************/
+
+package org.eclipse.hono.registry.infinispan;
+
+import org.eclipse.hono.deviceregistry.Application;
```
Comment (sophokles73) [inline services/infinispan-device-registry/src/main/java/org/eclipse/hono/registry/infinispan/InfinispanRegistry.java]: IMHO we should not have a dependency on another registry implementation.

File: services/infinispan-device-registry/src/main/java/org/eclipse/hono/registry/infinispan/InfinispanRegistryConfig.java [hunk 29]
```diff
@@ -0,0 +1,49 @@
...
+import java.io.IOException;
+
+/**
+ * Spring Boot configuration for the Device Registry application.
+ *
+ */
+@Configuration
+public class InfinispanRegistryConfig extends ApplicationConfig {
+
+    @Value("${infinispan.config.file}")
```
Comment (sophokles73) [inline services/infinispan-device-registry/src/main/java/org/eclipse/hono/registry/infinispan/InfinispanRegistryConfig.java]: I'd rather have this defined as a property in a subclass of `org.eclipse.hono.config.ApplicationConfigProperties` ...

File: services/infinispan-device-registry/src/main/resources/application.yml (hunk 26 above)
Comment (jbtrystram) [reply services/infinispan-device-registry/src/main/resources/application.yml]: I guess this is a legacy from the other device registry (which is configurable with either `dummy` or `file` ). 
But as we will probably package this as a separate image there is no reason to use this.
Maybe later we could use this to choose between the embedded cache or the client/server one ? @dejanb what do you think, should we keep that option open ?

Comment (dejanb) [reply services/infinispan-device-registry/src/main/resources/application.yml]: We can always add it later if we need it. Right now, as this registry has its own app we don't really need it

File: services/infinispan-device-registry/src/main/java/org/eclipse/hono/registry/infinispan/InfinispanRegistryConfig.java (hunk 27 above)
Comment (dejanb) [reply services/infinispan-device-registry/src/main/java/org/eclipse/hono/registry/infinispan/InfinispanRegistryConfig.java]: deviceregistry package contains a number of useful classes, not related only to file-based registry. We can move those to service-base module? Or move file-based registry into its own package (and the dummy one)?
As for the ApplicationConfig we can refactor it to remove all file-based registry specific stuff from it.

File: services/infinispan-device-registry/src/main/java/org/eclipse/hono/registry/infinispan/InfinispanRegistry.java (hunk 28 above)
Comment (dejanb) [reply services/infinispan-device-registry/src/main/java/org/eclipse/hono/registry/infinispan/InfinispanRegistry.java]: See above ... But in the case of Application class I don't see any file-based registry specific code in it.

File: services/infinispan-device-registry/src/main/java/org/eclipse/hono/registry/infinispan/InfinispanRegistryConfig.java (hunk 29 above)
Comment (dejanb) [reply services/infinispan-device-registry/src/main/java/org/eclipse/hono/registry/infinispan/InfinispanRegistryConfig.java]: Just to repeat that we I think there's a value in common config class for the registry implementations (that could extend ApplicationConfigProperties).

File: services/infinispan-device-registry/src/main/resources/application.yml (hunk 26 above)
Comment (jbtrystram) [reply services/infinispan-device-registry/src/main/resources/application.yml]: I removed it

File: deploy/hono-service-infinispan-device-registry-config.yml [hunk 30]
```diff
@@ -0,0 +1,29 @@
+hono:
+  app:
+    maxInstances: 1
+    healthCheckBindAddress: 0.0.0.0
+    healthCheckPort: 8088
+  auth:
+    host: hono-service-auth.hono
+    port: 5671
+    keyPath: /workspace/oss/hono/deploy/target/deploy/helm/hono-demo-certs-jar/device-registry-key.pem
```
Comment (jbtrystram) [inline deploy/hono-service-infinispan-device-registry-config.yml]: looks like tied to your machine :)

Comment (dejanb) [reply deploy/hono-service-infinispan-device-registry-config.yml]: Thanks for spotting it. It was a test file, shouldn't have been committed at all.

**Pull request closed without merging**